File: app/src/test/java/com/dpis/module/templates/ToolsWorkspaceBinderSourceSmokeTest.java
package com.dpis.module;

import com.dpis.module.diagnostics.presentation.LogGate;

import static org.junit.Assert.assertTrue;
import static org.junit.Assert.assertFalse;

import java.io.IOException;

import org.junit.Test;

public class ToolsWorkspaceBinderSourceSmokeTest {
    @Test
    public void mainActivityWiresToolsWorkspaceBinderLifecycle() throws IOException {
        String source = read("src/main/java/com/dpis/module/MainActivity.java");
        String workspace = read("src/main/java/com/dpis/module/settings/presentation/ToolsWorkspace.kt");

        assertTrue(source.contains("private ToolsWorkspace toolsWorkspace;"));
        assertTrue(source.contains("toolsWorkspace = new ToolsWorkspace("));
        assertTrue(workspace.contains("private val binder = ToolsWorkspaceBinder("));
        assertTrue(workspace.contains(
                "WindowInsetsBinder.applySystemBarPadding(toolbar, false, true, false, false)"));
        assertTrue(workspace.contains("TouchFeedbackBinder.bindPressHaptic(view)"));
        assertTrue(workspace.contains("LogGate.ensureEnabled("));
        assertTrue(source.contains("bindToolsWorkspace();"));
        assertTrue(source.contains("toolsWorkspace.onStart();"));
        assertTrue(source.contains("toolsWorkspace.onResume();"));
        assertTrue(source.contains("toolsWorkspace.onStop();"));
        assertTrue(source.contains("toolsWorkspace.onActivityResult(requestCode, resultCode, data);"));
    }

    @Test
    public void toolsWorkspaceBinderOwnsSystemFontScaleToolBinder() throws IOException {
        String source = read("src/main/java/com/dpis/module/settings/presentation/ToolsWorkspaceBinder.kt");

        assertTrue(source.contains("private var fontScaleToolBinder: SystemFontScaleToolBinder? = null"));
        assertTrue(source.contains(
                "fontScaleToolBinder = SystemFontScaleToolBinder(host.activity(), workspaceView, host)"));
        assertTrue(source.contains("fontScaleToolBinder?.refreshFromSystem()"));
        assertTrue(source.contains("fontScaleToolBinder?.collapseAndRefreshFromSystem()"));
    }

    @Test
    public void systemFontScalePermissionPanelOwnsAuthorizationClick() throws IOException {
        String source = read("src/main/java/com/dpis/module/settings/presentation/SystemFontScaleToolBinder.kt");

        assertTrue(source.contains("host.bindPressHaptic(permissionOverlay)"));
        assertTrue(source.contains("permissionOverlay?.setOnClickListener { openWriteSettingsPermission() }"));
        assertTrue(source.contains("setVisible(operationGroup, expanded && (current.canWrite || current.unavailable))"));
    }

    private static String read(String relativePath) throws IOException {
        return SourceSmokeTestPaths.read(relativePath);
    }
}

File: app/src/test/java/com/dpis/module/templates/ToolsWorkspaceBinderSourceSmokeTest.kt
package com.dpis.module

import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class ToolsWorkspaceBinderSourceSmokeTest {
    @Test
    fun mainActivityWiresToolsWorkspaceLifecycle() {
        val startup = read(
            "src/main/java/com/dpis/module/ui/presentation/MainStartupSession.kt"
        )
        val workspaceSession = read(
            "src/main/java/com/dpis/module/ui/presentation/MainWorkspaceSession.kt"
        )
        val hostWiring = read(
            "src/main/java/com/dpis/module/ui/presentation/MainHostWiringSession.kt"
        )
        val workspace = read("src/main/java/com/dpis/module/settings/presentation/ToolsWorkspace.kt")

        assertTrue(hostWiring.contains("var toolsWorkspace: ToolsWorkspace?"))
        assertTrue(hostWiring.contains("toolsWorkspace = ToolsWorkspace("))
        assertTrue(workspace.contains("SystemFontScaleToolPresenter("))
        assertFalse(workspace.contains("ToolsWorkspaceBinder"))
        assertTrue(workspaceSession.contains("fun bindToolsWorkspace("))
        assertTrue(startup.contains("mainWorkspaceSession.bindForLifecycle("))
        assertTrue(startup.contains("hostWiringSession.toolsWorkspace?.onStart()"))
        assertTrue(startup.contains("hostWiringSession.toolsWorkspace?.onResume()"))
        assertTrue(startup.contains("hostWiringSession.toolsWorkspace?.onStop()"))
        assertTrue(
            startup.contains(
                "hostWiringSession.toolsWorkspace?.onActivityResult(requestCode, resultCode, data)",
            )
        )
    }

    @Test
    fun composeToolsWorkspaceOwnsFontScalePermissionAndApply() {
        val content = read(
            "src/main/java/com/dpis/module/tools/presentation/ToolsWorkspaceContent.kt"
        )
        val workspace = read("src/main/java/com/dpis/module/settings/presentation/ToolsWorkspace.kt")

        assertTrue(content.contains("onRequestPermission"))
        assertTrue(content.contains("R.string.system_font_scale_apply"))
        assertTrue(workspace.contains("fun requestPermission()"))
        assertTrue(workspace.contains("Settings.ACTION_MANAGE_WRITE_SETTINGS"))
        assertTrue(workspace.contains("presenter.apply()"))
    }

    private fun read(relativePath: String): String = SourceSmokeTestPaths.read(relativePath)
}
